### PR TITLE
Chore: rebuild mimir-build-image when dockerfile change

### PR DESCRIPTION
**Pull request by ying-jeanne**
#### What this PR does
This PR would detect change in mimir-build-image and build/push a PR when it is needed. It would update the current PR with new image tag in makefile after update docker image.
#### Which issue(s) this PR fixes or relates to

Fixes #<issue number>

#### Checklist

- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by ying-jeanne** [comment]
@mimir-build-image

**Comment by ying-jeanne** [comment]
@mimir-build-image

**Comment by ying-jeanne** [comment]
@mimir-build-image

**Comment by ying-jeanne** [comment]
@mimir-build-image

**Comment by ying-jeanne** [comment]
@mimir-build-image

File: .github/workflows/push-mimir-build-image.yml
Comment (charleskorn) [inline .github/workflows/push-mimir-build-image.yml:86]: If we are going to compute the image tag by taking a checksum of the build image Dockerfile, should we just do this in the Makefile when determining the image tag? If we do it in the Makefile, then it removes the need to modify the Makefile every time the build image changes.

File: .github/workflows/push-mimir-build-image.yml
Comment (charleskorn) [inline .github/workflows/push-mimir-build-image.yml:90]: Should we use a command-line tool like [`skopeo`](https://github.com/containers/skopeo) to check if the image exists? We use this in other (non-public) repositories for this exact purpose, and it supports authentication when talking to the Docker Hub API, which would help avoid rate limiting as well.

File: docs/internal/how-to-update-the-build-image.md
Comment (charleskorn) [inline docs/internal/how-to-update-the-build-image.md:12]: Protection like this likely won't work as anyone who wants to do something malicious could modify the workflow to remove this requirement as part of their PR. 

I think we instead need to use GitHub's [workflow approval option for outside collaborators](https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks). We currently have this set to "require approval for first-time contributors", but should probably change this to "require approval for all outside collaborators".

File: .github/workflows/push-mimir-build-image.yml
Comment (pstibrany) [reply .github/workflows/push-mimir-build-image.yml:86]: What you describe was the original intent, but there's a problem... if PR modifies build image Dockerfile, how will Makefile find *previous* build image to actually run actions with? New build image doesn't exist yet, so it can't be used.

File: .github/workflows/push-mimir-build-image.yml
Comment (ying-jeanne) [reply .github/workflows/push-mimir-build-image.yml:90]: yes, you are right, Peter raised same concerns, I plan to change it before putting it to ready to review.

File: docs/internal/how-to-update-the-build-image.md
Comment (ying-jeanne) [reply docs/internal/how-to-update-the-build-image.md:12]: agreed, it would simplify also the workflow itself.

**Comment by ying-jeanne** [comment]
This PR would be merged when the permission for github action run is configured as discussed [here](https://github.com/grafana/mimir/pull/5335#discussion_r1241402161)

File: .github/workflows/push-mimir-build-image.yml
Comment (charleskorn) [reply .github/workflows/push-mimir-build-image.yml]: If we do these steps (checking that the build image has been built and rebuilding it if it hasn't) at the start of the `prepare` job in the existing CI workflow, then all later steps can use the new image. 

Then, the only steps that need to run without the build image are computing the build image tag (which has to run outside the build image always anyway), and checking if the build image exists.

Comment (pstibrany) [reply .github/workflows/push-mimir-build-image.yml]: 👍

Comment (pstibrany) [reply .github/workflows/push-mimir-build-image.yml]: We were hoping to include additional information in the image tag: PR number that last changed the docker-image (to provide some trace where the Dockerfile update came from), and Go version included in the image. When computing the tag, including Go version would be easy (grep dockerfile), but PR number would be tricky.

Comment (pstibrany) [reply .github/workflows/push-mimir-build-image.yml]: One more scenario came up in discussion: what if there are two PRs updating build-image at the same time. When computing the tag, they both could get merged, and then final integration of the two changes would happen on the `main` branch -- when running "prepare" job again.

If we keep reference to the tag instead, then after merging one PR, second would have a conflict and merge would be blocked until conflict is resolved. That would move integration of the two changes into the PR.

(In practice we don't update build image that often.)

File: docs/internal/how-to-update-the-build-image.md
Comment (pstibrany) [reply docs/internal/how-to-update-the-build-image.md]: Repository has been configured to require approval for all runs of GHA on PRs from outside collaborators.

File: .github/workflows/push-mimir-build-image.yml
Comment (pstibrany) [inline .github/workflows/push-mimir-build-image.yml:6]: I'm curious why does this workflow need run again when PR is marked as "Ready for Review"?

File: .github/workflows/push-mimir-build-image.yml
Comment (pstibrany) [inline .github/workflows/push-mimir-build-image.yml:54]: nit: `skopeo inspect --raw "docker://grafana/mimir:latest"` seems to work too (ie. without `docker.io/` prefix). I'd suggest to use that to avoid hardcoding `docker.io` if it's not required.

File: .github/workflows/push-mimir-build-image.yml
Comment (pstibrany) [inline .github/workflows/push-mimir-build-image.yml:39]: We should reuse image name from Makefile. (BUILD_IMAGE)

File: .github/workflows/push-mimir-build-image.yml
Comment (pstibrany) [inline .github/workflows/push-mimir-build-image.yml:67]: question: why PR_URL and not just PR number, or even no extra PR number argument? Are there any benefits?

File: .github/workflows/push-mimir-build-image.yml
Comment (pstibrany) [inline .github/workflows/push-mimir-build-image.yml:57]: Let's include tag in the message here and in `else` condition.

File: .github/workflows/push-mimir-build-image.yml
Comment (pstibrany) [inline .github/workflows/push-mimir-build-image.yml:72]:
```suggestion
            gh pr comment $PR_URL --body "**Not building new version of mimir-build image, image <insert image name here> already exits.**"
```

File: .github/workflows/push-mimir-build-image.yml
Comment (pstibrany) [inline .github/workflows/push-mimir-build-image.yml:68]: I would modify the message to active message: "Building new version of mimir-build-image. Image will be pushed to docker hub and commit will be added to the PR with new image version".

(I would also not mention that "This PR" triggers something... as more commits to the PR can also trigger that.)

```suggestion
            gh pr comment $PR_URL --body "**Building new version of mimir-build image**. After image is built and pushed to docker hub, new commit will be added to this PR with new image version <insert image tag here>."
```

File: .github/workflows/push-mimir-build-image.yml
Comment (pstibrany) [inline .github/workflows/push-mimir-build-image.yml:87]:
```suggestion
          main_version=$(make print-build-image | cut -d ':' -f 2)
```

**Review by pstibrany — COMMENTED** [review]
Thank you, nice work. I've left some minor comments, but this looks good.

File: docs/internal/contributing/how-to-upgrade-golang-version.md
Comment (pstibrany) [inline docs/internal/contributing/how-to-upgrade-golang-version.md:6]: Opening PR will already trigger building of the image.

File: .github/workflows/push-mimir-build-image.yml
Comment (pstibrany) [inline .github/workflows/push-mimir-build-image.yml:47]: Can we please include "pr" prefix before PR number? eg. `pr5335-d9796228c4f26ab68a87f048f39245de`.

Nit: Also perhaps we don't need full md5 hash? (`awk '{print substr($1, 0, 10)}'`), so the end result would be: `pr5335-c90da5e26b`. WDYT?

File: Makefile
Comment (pstibrany) [inline Makefile:185]: Do we need to update image in this PR? We should revert this change and any changes in Dockerfile.

File: .github/workflows/push-mimir-build-image.yml
Comment (ying-jeanne) [reply .github/workflows/push-mimir-build-image.yml:6]: I was thinking when someone opened PR and set to ready-for-review later, but in that case it would already have the image up to date. we can remove it.

File: .github/workflows/push-mimir-build-image.yml
Comment (ying-jeanne) [reply .github/workflows/push-mimir-build-image.yml:67]: they are the same for pr in same repo, replacing by number instead.

File: Makefile
Comment (ying-jeanne) [reply Makefile:185]: no, this is auto-updated by workflow. agreed.

Comment (ying-jeanne) [reply Makefile]: I will remove the newline i added in the dockerfile, if there is no change detected, the workflow would not run anymore.

File: .github/workflows/push-mimir-build-image.yml
Comment (charleskorn) [inline .github/workflows/push-mimir-build-image.yml:8]: Should this be `mimir-build-image/*`, in case we want to be able to add more files to the image? If so, we'll need to adjust the checksum calculation below as well.

File: .github/workflows/push-mimir-build-image.yml
Comment (charleskorn) [inline .github/workflows/push-mimir-build-image.yml:13]: This comment seems outdated.

File: .github/workflows/push-mimir-build-image.yml
Comment (charleskorn) [inline .github/workflows/push-mimir-build-image.yml:75]: [nit] clarify messages

```suggestion
          if [ ${{ steps.check_build.outputs.build }} == 'true' ]; then
           gh pr comment $PR_URL --body "**Building new version of `mimir-build-image`**. After image is built and pushed to Docker Hub, \
           a new commit will automatically be added to this PR with new image version `$IMAGE:$TAG`. This can take up to 1 hour."
          else
            echo "This PR will not trigger a build of mimir-build-image"
            gh pr comment $PR_URL --body "**Not building new version of `mimir-build image`**. This PR modifies `mimir-build-image/Dockerfile`, but the image `$IMAGE:$TAG` already exists."
          fi
```

File: .github/workflows/push-mimir-build-image.yml
Comment (charleskorn) [inline .github/workflows/push-mimir-build-image.yml:1]: Should we configure this workflow to only allow one build at a time, and cancel in-progress builds when new commits are pushed? ([docs](https://docs.github.com/en/actions/using-jobs/using-concurrency))

File: .github/workflows/push-mimir-build-image.yml
Comment (charleskorn) [inline .github/workflows/push-mimir-build-image.yml:3]: I don't think this workflow can be triggered manually - don't we need a `workflow_dispatch` trigger?

File: docs/internal/contributing/how-to-upgrade-golang-version.md
Comment (charleskorn) [inline docs/internal/contributing/how-to-upgrade-golang-version.md:6]: [nit]
```suggestion
2. Open PR and wait for mimir-build-image to be updated. Refer to the [documentation](https://github.com/grafana/mimir/blob/main/docs/internal/how-to-update-the-build-image.md) for more information.
```

File: docs/internal/how-to-update-the-build-image.md
Comment (charleskorn) [inline docs/internal/how-to-update-the-build-image.md:14]: Should we move this part to the top of the file, given it's the preferred procedure?

**Review by pstibrany — APPROVED** [review]
Thank you. I think this will be ready, once Charles' comments are addressed.

File: .github/workflows/push-mimir-build-image.yml
Comment (ying-jeanne) [reply .github/workflows/push-mimir-build-image.yml:8]: For now we have only the case for dockerfile (go, alpine update), wdyt let's try this first, if we see include repo is necessary, we can change that easily.

File: .github/workflows/push-mimir-build-image.yml
Comment (ying-jeanne) [reply .github/workflows/push-mimir-build-image.yml:1]: i don't know why i thought it is the default behavior, gonna update

File: .github/workflows/push-mimir-build-image.yml
Comment (charleskorn) [reply .github/workflows/push-mimir-build-image.yml:8]: Sure, sounds good to me.

**Review by charleskorn — APPROVED** [review]

**Pull request merged**
